File: next-app/src/lib/library/episodeOffset.ts
// The one rule for translating between the two episode-number spaces.
//
// TWO SPACES, AND THE ONLY PLACE THE CONVERSION IS STATED
//
//   local  the number on disk. Release groups routinely number continuously
//          across seasons, so the finale of a 10-episode second season whose
//          predecessor ran 28 is called 38.
//   site   the number the season itself uses, 1..episodeCount. This is what
//          `subscriptions/validate.go` range-checks, and what a reader means
//          when they say "episode 10".
//
// They were never reconciled. The grid inferred a shift from the files alone
// ("the lowest one I hold must be episode 1"), which is right for a full
// season and wrong for a tail — a lone finale numbered 38 rendered in slot 1.
// The watch push did not translate at all, so it sent 38 into a range check
// that stops at 10 and got a 400 every time; that season's progress had never
// synced, silently, for as long as the file had been in the library.
//
// The fix needs the same decision in both places, and the two live in
// different layers — `_services/episodeGridModel` renders, `lib/watchSync`
// pushes. Two copies of a rule this quiet is how they drift: the display
// would show episode 10 while the push sent 38, and nothing would fail. So it
// is stated once, here, in `lib/` where both may import it.
//
// The offset itself comes from GET /api/anime/{id}/episode-offset, which
// walks PREQUEL edges. See `resolveSeriesBinding.fetchEpisodeOffset`.

/**
 * A measured offset, or `undefined` when none was measured.
 *
 * `0` and `undefined` are OPPOSITE instructions, never two flavours of empty:
 * 0 says "nothing precedes this season, stop inferring a shift", undefined
 * says "we do not know, carry on inferring one". Any `offset ?? 0` at a call
 * site collapses the second into the first and renumbers a library against an
 * origin nobody established.
 */
export function readEpisodeOffset(value: unknown): number | undefined {
  return typeof value === "number" && Number.isInteger(value) && value >= 0
    ? value
    : undefined;
}

/**
 * Does this offset actually describe these files?
 *
 * The server knows the franchise; only the caller knows what is on disk, and
 * `format` is a coarse proxy for "counts toward the continuous numbering". So
 * a measurement is checked against the evidence before it is trusted: every
 * episode must land inside the season it claims to belong to.
 *
 * Returns false for an unmeasured offset, an unknown total, or an empty set —
 * "cannot be applied" rather than "declined", because the caller's next move
 * is the same in all three cases.
 */
export function offsetApplies(
  numbers: readonly number[],
  total: number | undefined,
  offset: number | undefined,
): boolean {
  if (offset === undefined || offset <= 0) return false;
  if (typeof total !== "number" || !Number.isInteger(total) || total <= 0) return false;
  if (numbers.length === 0) return false;
  return numbers.every((n) => {
    const site = n - offset;
    return site >= 1 && site <= total;
  });
}

/**
 * Local episode numbers → the numbers the site uses, or `null` when the
 * offset must not be applied.
 *
 * ALL OR NOTHING, and that is the load-bearing part. Translating the episodes
 * that happen to fit and leaving the rest alone would put two numbering
 * spaces inside one request — and worse, it would let the grid and the push
 * disagree about the same file, because the grid applies `offsetApplies` to
 * the whole card. `null` means "send what you already had", which is exactly
 * what the grid means by rendering the stored number.
 */
export function toSiteEpisodes(
  numbers: readonly number[],
  total: number | undefined,
  offset: number | undefined,
): number[] | null {
  if (!offsetApplies(numbers, total, offset)) return null;
  const shift = offset as number;
  return numbers.map((n) => n - shift);
}

File: go-api/internal/anime/episode_offset.go
// episode_offset.go — GET /api/anime/{anilistId}/episode-offset
//
// How many episodes precede a season in its franchise's continuous
// numbering. Release groups routinely number across seasons: a file holding
// the finale of a 10-episode second season is called 38 when the first
// season ran 28. Nothing in the library knew that, and two things broke on
// it:
//
//   - the grid inferred the shift as "the lowest file I hold is episode 1",
//     which is only true for a user who has the season's opener. A user
//     holding just the finale saw it rendered in slot 1.
//   - the watch push sent the stored 38, which subscriptions/validate.go
//     range-checks against anime_cache.episodes and rejects with 400. That
//     season's progress had therefore never synced, silently.
//
// Both wanted one number, and the database already had everything needed to
// derive it — `anime_relations` has carried PREQUEL edges since 0001, filled
// from the `relations` block the AniList detail query already asks for. See
// GetAbsoluteEpisodeOffset in queries/anime_cache.sql for the walk.
//
// Its own endpoint rather than a field on /search or /{anilistId}: the
// library reads this once per bound series and stores it, so it does not
// belong on either of the two hot paths that would then compute it for every
// row of every result.

package anime

import (
	"context"
	"errors"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"github.com/jackc/pgx/v5"

	dbgen "github.com/lawrenceli0228/animego/go-api/internal/db/gen"
	"github.com/lawrenceli0228/animego/go-api/internal/httpx"
)

// episodeOffsetResponse is deliberately two fields rather than a nullable
// number.
//
// "nothing precedes this season" and "we could not work out what precedes
// this season" are different answers, and a single `offset: number | null`
// invites the one line of client code — `offset ?? 0` — that turns the
// second into the first. Renumbering a user's files against an origin
// nobody established is exactly the failure this endpoint exists to stop, so
// the caller has to read `known` before it can reach a number at all.
type episodeOffsetResponse struct {
	Known  bool  `json:"known"`
	Offset int32 `json:"offset"`
}

// EpisodeOffset serves the absolute episode offset for one anime.
//
// Unknown is a 200 with `known:false`, never a 404 or a 500. It is the
// ordinary answer for roughly a quarter of the catalogue — relation rows are
// only written for anime whose detail has been fetched — and a client that
// had to distinguish "unknown" from "request failed" by status code would
// end up retrying a stable answer forever.
func EpisodeOffset(q dbgen.Querier) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		ctx, cancel := context.WithTimeout(req.Context(), queryTimeout)
		defer cancel()

		raw := chi.URLParam(req, "anilistId")
		id, err := strconv.Atoi(raw)
		if err != nil || id <= 0 {
			httpx.Fail(w, httpx.NewError(
				http.StatusBadRequest,
				httpx.CodeValidationError,
				"无效的番剧 ID",
			))
			return
		}

		row, err := q.GetAbsoluteEpisodeOffset(ctx, int32(id))
		if err != nil {
			// No row means the anchor itself is not cached, so there is no
			// chain to walk. That is an unknown offset, not a failure: the
			// id is perfectly valid and the client's answer is the same one
			// it gets for a chain that runs off the end of the cache.
			if errors.Is(err, pgx.ErrNoRows) {
				httpx.Data(w, http.StatusOK, episodeOffsetResponse{Known: false})
				return
			}
			httpx.Fail(w, httpx.WrapError(err, http.StatusInternalServerError, httpx.CodeServerError, "query failed"))
			return
		}

		// `known` is a *bool because the SQL expression is nullable in
		// principle. A nil here is not "unknown-ish" — it would mean the
		// walk produced a row whose completeness could not be evaluated, so
		// it collapses to the same refusal rather than to a default.
		known := row.Known != nil && *row.Known
		out := episodeOffsetResponse{Known: known}
		if known {
			out.Offset = row.AbsoluteOffset
		}
		httpx.Data(w, http.StatusOK, out)
	}
}

// episodeOffsetItem is one row of the batch answer. `known` carries the same
// meaning it does on the single-id endpoint, and for the same reason.
type episodeOffsetItem struct {
	AnilistID int32 `json:"anilistId"`
	Known     bool  `json:"known"`
	Offset    int32 `json:"offset"`
}

// EpisodeOffsets serves offsets for many anime at once.
//
// It exists because the library's one-shot backfill cannot use the per-id
// endpoint: a 200-series library would make 200 requests against a 1 req/s
// per-IP bucket. `/api/anime/episodes` already solved that for episode
// counts, so this mirrors it — same id parsing, same cap, same cache policy,
// same `{data: [...]}` envelope — rather than inventing a second shape for
// the same problem.
//
// An id that is not in the cache produces NO ITEM rather than an item saying
// unknown. Absent and known:false mean the same thing to the caller, and the
// episode-count sweep already treats a missing row that way.
func EpisodeOffsets(q dbgen.Querier) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		ctx, cancel := context.WithTimeout(req.Context(), queryTimeout)
		defer cancel()

		ids, err := parseAnilistIDList(req.URL.Query().Get("ids"), maxEpisodeIDs)
		if err != nil {
			httpx.Fail(w, httpx.NewError(
				http.StatusBadRequest,
				httpx.CodeValidationError,
				err.Error(),
			))
			return
		}

		rows, queryErr := q.GetAbsoluteEpisodeOffsets(ctx, ids)
		if queryErr != nil {
			httpx.Fail(w, httpx.WrapError(queryErr, http.StatusInternalServerError, httpx.CodeServerError, "query failed"))
			return
		}

		items := make([]episodeOffsetItem, 0, len(rows))
		for _, r := range rows {
			known := r.Known != nil && *r.Known
			item := episodeOffsetItem{AnilistID: r.AnilistID, Known: known}
			if known {
				item.Offset = r.AbsoluteOffset
			}
			items = append(items, item)
		}

		// The same policy the episode counts use. An offset is a fact about a
		// franchise's shape, which changes even less often than an episode
		// count does.
		w.Header().Set("Cache-Control", episodeCountsCacheControl)
		httpx.Data(w, http.StatusOK, items)
	}
}
